src/Test/Utilities/Shared/Syntax/SourceUtilities.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+using Microsoft.CodeAnalysis.Text;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Roslyn.Test.Utilities.Syntax
+{
+    internal sealed class SegmentedSourceText: SourceText
+    {
+        const int BLOCK_SIZE = 1024;
+
+        List<char[]> blocks = new List<char[]>();
+
+        public SegmentedSourceText(IEnumerable<char> from)
+        {
+            blocks.Add(new char[BLOCK_SIZE]);
+            var i = 0;
+
+            foreach (char ch in from)
+            {
+                if (i == BLOCK_SIZE)
+                {
+                    i = 0;
+                    blocks.Add(new char[BLOCK_SIZE]);
+                }
+                blocks.Last()[i] = ch;
+                i++;
+            }
+        }
+
+        public override char this[int position]
+        {
+            get
+            {
+                var block = position / BLOCK_SIZE;
+                var offset = position % BLOCK_SIZE;
+                return blocks[block][offset];
```

Comment:
It looks like an assumption is made that the last block is completely filled with data? What this assumption is based on?